File: src/java/com/springbok/station/EarthStation.java
/* Copyright (C) 2022 Springbok LLC

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or (at
your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
package com.springbok.station;

import Jama.Matrix;
import com.springbok.antenna.Antenna;
import com.springbok.twobody.Coordinates;
import com.springbok.twobody.EarthConstants;
import com.springbok.twobody.ModJulianDate;
import com.springbok.utility.MException;

import java.io.Serializable;

/**
 * Describes the position of a ground based sensor.
 *
 * @author Raymond LeClair
 */
@SuppressWarnings("serial")
public class EarthStation extends Station implements Serializable {

    // Describes a space or an Earth station beam.
    private Beam beam;
    // Geodetic Latitude [rad]
    protected double varphi;
    // Longitude [rad]
    protected double lambda;
    // Flag indicating whether to do multiplexing, or not
    private boolean doMultiplexing;
    // Date number at which the position vector occurs
    private ModJulianDate dNm;
    // Geocentric equatorial inertial position vector [er]
    private Matrix r_gei;

    // Geocentric equatorial rotating position [er]
    protected Matrix r_ger;

    /**
     * Constructs an EarthStation.
     *
     * @param sensorId Identifier
     * @param varphi   Geodetic Latitude [rad]
     * @param lambda   Longitude [rad]
     */
    public EarthStation(String sensorId, double varphi, double lambda) {

        // Fundamental values.
        super(sensorId);
        this.varphi = varphi;
        this.lambda = lambda;

        // Derived values.
        this.r_ger = compute_r_ger(null);
    }

    /**
     * Constructs an EarthStation.
     *
     * @param sensorId Identifier
     * @param varphi   Geodetic Latitude [rad]
     * @param lambda   Longitude [rad]
     */
    public EarthStation(String sensorId, Antenna transmitAntenna, Antenna receiveAntenna, Emission emission,
                        Beam beam, double varphi, double lambda, boolean doMultiplexing) {
        // Fundamental values.
        super(sensorId, transmitAntenna.copy(), receiveAntenna.copy(), emission.copy());
        this.varphi = varphi;
        this.lambda = lambda;
        this.set_beam(beam);
        this.doMultiplexing = doMultiplexing;

        // Derived values.
        this.r_ger = compute_r_ger(null);
    }

    /**
     * Constructs an EarthStation.
     */
    public EarthStation() {
        super();
    }

    /**
     * Copies an EarthStation.
     *
     * @return A new EarthStation instance
     */
    public EarthStation copy() {
        EarthStation that = new EarthStation(this.get_stationId(), this.get_transmitAntenna().copy(), this.get_receiveAntenna().copy(),
                this.get_emission().copy(), this.beam.copy(), this.varphi, this.lambda, this.doMultiplexing);
        that.compute_r_gei(that.dNm);

        return that;
    }

    /**
     * Sets geodetic latitude [rad].
     *
     * @param varphi The geodetic latitude [rad]
     */
    public void set_varphi(double varphi) {
        this.varphi = varphi;
        this.r_ger = compute_r_ger(null);
    }

    /**
     * Gets geodetic latitude [rad].
     *
     * @return The geodetic latitude [rad]
     */
    public double get_varphi() {
        return varphi;
    }

    /**
     * Sets longitude [rad]
     *
     * @param lambda The longitude [rad]
     */
    public void set_lambda(double lambda) {
        this.lambda = lambda;
        this.r_ger = compute_r_ger(null);
    }

    /**
     * Gets longitude [rad]
     *
     * @return The longitude [rad]
     */
    public double get_lambda() {
        return lambda;
    }

    /**
     * Sets the Beam object for the Earth station.
     *
     * @param beam The Beam object to set
     * @throws MException if the multiplicity of the Beam object is not equal to 1
     */
    public void set_beam(Beam beam) {
        if (beam.get_multiplicity() != 1) {
            throw new MException("Springbok:IllegalArgumentException",
                    "An Earth station beam must have multiplicity one");
        }
        this.beam = beam;
        this.beam.assign(this.doMultiplexing);
    }

    /**
     * Gets the beam object.
     *
     * @return The beam object
     */
    public Beam get_beam() {
        return beam;
    }

    /**
     * Sets the flag indicating whether to perform multiplexing or not.
     *
     * @param bool Flag indicating if to perform multiplexing, or not
     */
    public void set_doMultiplexing(boolean bool) {
        doMultiplexing = bool;
    }

    /**
     * Checks if multiplexing is enabled.
     *
     * @return true if multiplexing is enabled, false otherwise.
     */
    public boolean doMultiplexing() {
        return this.doMultiplexing;
    }

    /**
     * Computes geocentric equatorial inertial position vector.
     *
     * @param dNm Date number at which the position vector occurs
     * @return Geocentric equatorial inertial position vector [er]
     */
    public Matrix compute_r_gei(ModJulianDate dNm) {
        if (dNm != null && this.dNm != null && !dNm.equals(this.dNm)) {
            this.dNm = dNm;
            this.r_gei = Coordinates.ger2gei(this.r_ger, dNm);
        }
        return this.r_gei;
    }

    /**
     * Computes the geocentric equatorial rotating position vector.
     *
     * @param dNm Date number at which the position vector occurs
     * @return Geocentric equatorial rotating position vector [er]
     */
    public Matrix compute_r_ger(ModJulianDate dNm) {
        if (this.r_ger == null) {
            double N = 1.0 / (Math.sqrt(1 - EarthConstants.f * (2 - EarthConstants.f) * Math.pow(Math.sin(varphi), 2)));
            double h = 0.0;
            double[][] elements = {{(N + h) * Math.cos(varphi) * Math.cos(lambda)},
                    {(N + h) * Math.cos(varphi) * Math.sin(lambda)},
                    {(Math.pow(1.0 - EarthConstants.f, 2) * N + h) * Math.sin(varphi)}};
            this.r_ger = new Matrix(elements);
        }
        return this.r_ger;
    }
}

File: src/matlab/com/springbok/station/EarthStation.m
% Copyright (C) 2022 Springbok LLC
% 
% This program is free software: you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation, either version 3 of the License, or (at
% your option) any later version.
% 
% This program is distributed in the hope that it will be useful, but
% WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
% General Public License for more details.
% 
% You should have received a copy of the GNU General Public License
% along with this program.  If not, see <https://www.gnu.org/licenses/>.
% 
% TODO: Add altitude
classdef EarthStation < Station
% Describes an Earth station
  
  properties (SetAccess = private, GetAccess = public)
    
    % A beam
    beam

    % Geodetic latitude [rad]
    varphi
    % Longitude [rad]
    lambda
    % Altitude [er]
    h

    % Flag indicating whether to do multiplexing, or not
    doMultiplexing
    
    % Geocentric equatorial rotating position [er]
    r_ger

    % Date number at which the position vector occurs
    dNm
    % Geocentric equatorial inertial position vector [er]
    r_gei
    
  end % properties (SetAccess = private, GetAccess = public)
  
  methods
    
    function this = EarthStation(stationId, transmitAntenna, ...
                                 receiveAntenna, emission, beam, ...
                                 varphi, lambda, doMultiplexing)
      % Constructs an EarthStation.
      %
      % Parameters
      %   stationId - Identifier for station
      %   transmitAntenna - Antenna gain, and pattern
      %   receiveAntenna - Antenna gain, pattern, and noise
      %     temperature
      %   emission - Signal power, frequency, and requirement
      %   beam - A beam
      %   varphi - Geodetic latitude [rad]
      %   lambda - Longitude [rad]
      %   doMultiplexing - Flag indicating whether to do multiplexing,
      %     or not
      
      % Invoke the superclass constructor
      if nargin == 0
        superArgs = {};
        
      else
        superArgs{1} = stationId;
        superArgs{2} = transmitAntenna;
        superArgs{3} = receiveAntenna;
        superArgs{4} = emission;
        
      end % if
      this@Station(superArgs{:});
      if nargin == 0
        return

      end % if
      
      % Assign properties
      this.set_transmitAntenna(transmitAntenna);
      this.set_receiveAntenna(receiveAntenna);
      this.set_beam(beam);
      this.set_varphi(varphi);
      this.set_lambda(lambda);
      this.set_h(0.0);
      this.set_doMultiplexing(doMultiplexing)
      
    end % EarthStation()
    
    function that = copy(this)
    % Copies an EarthStation.
    %
    % Returns
    %   that - A new EarthStation instance
      that = EarthStation(this.stationId, this.transmitAntenna.copy(), ...
                          this.receiveAntenna.copy(), this.emission.copy(), ...
                          this.beam.copy(), this.varphi, this.lambda, ...
                          this.doMultiplexing);
      that.set_h(this.h);
      that.compute_r_gei(this.dNm);

    end % copy()

    function set_transmitAntenna(this, transmitAntenna)
    % Sets the transmit antenna.
    %
    % Parameters
    %   transmitAntenna - Transmit antenna
      if ~isa(transmitAntenna.pattern, 'EarthPattern')
        MEx = MException('Springbok:IllegalArgumentException', ...
                         'The transmit antenna must have an Earth pattern');
        throw(MEx);

      end % if
      set_transmitAntenna@Station(this, transmitAntenna);
      
    end % set_transmitAntenna()
    
    function set_receiveAntenna(this, receiveAntenna)
    % Sets the receive antenna.
    %
    % Parameters
    %   receiveAntenna - Receive antenna
      if ~isa(receiveAntenna.pattern, 'EarthPattern')
        MEx = MException('Springbok:IllegalArgumentException', ...
                         'The receive antenna must have an Earth pattern');
        throw(MEx);

      end % if
      set_receiveAntenna@Station(this, receiveAntenna);
      
    end % set_receiveAntenna()

    function set_beam(this, beam)
    % Sets beam.
    %
    % Parameters
    %   beam - A beam
      if ~isa(beam, 'Beam')
        MEx = MException('Springbok:IllegalArgumentException', ...
                         'Input must be of class "Beam"');
        throw(MEx);

      elseif length(beam) > 1
        MEx = MException('Springbok:IllegalArgumentException', ...
                         'An Earth station must have a single beam');
        throw(MEx);

      elseif beam.multiplicity ~= 1
        MEx = MException('Springbok:IllegalArgumentException', ...
                         'An Earth station beam must have multiplicity one');
        throw(MEx);

      end % if
      this.beam = beam;
      this.beam.assign(this.doMultiplexing);
      
    end % set_beam()
    
    function set_varphi(this, varphi)
    % Sets geodetic latitude [rad].
    % 
    % Parameters
    %   varphi - The geodetic latitude [rad]
      this.varphi = TypeUtility.set_numeric(varphi);
      this.compute_r_ger();
      
    end % set_varphi()
    
    function set_lambda(this, lambda)
    % Sets longitude [rad]
    % 
    % Parameters
    %   lambda - The longitude [rad]
      this.lambda = TypeUtility.set_numeric(lambda);
      this.compute_r_ger();

    end % set_lambda()

    function set_h(this, h)
    % Sets altitude [er].
    % 
    % Parameters
    %   h - The altitude [er]
      this.h = TypeUtility.set_numeric(h);
      this.compute_r_ger();

    end % set_lambda()

    function set_doMultiplexing(this, doMultiplexing)
    % Sets flag indicating whether to do multiplexing, or not.
    %
    % Parameters
    %   doMultiplexing - Flag indicating whether to do
    %     multiplexing, or not.
      this.doMultiplexing = TypeUtility.set_logical(doMultiplexing);

    end % if

    function r_ger = compute_r_ger(this, varargin)
    % Computes the geocentric equatorial rotating position
    % vector [er]. (MG-5.83)
      if isempty(this.varphi) || isempty(this.lambda) || isempty(this.h)
        return

      end % if
      if nargin == 1
        r_glla = [this.varphi; ...
                  this.lambda; ...
                  this.h];
        this.r_ger = Coordinates.glla2ger(r_glla);
      else
        % Anticipate being called with an additional date number
        % argument for consistency with the EarthStationInMotion and
        % SpaceStation classes. If so, the geocentric equatorial
        % rotating position vector will have been computed during
        % construction.
      end % if
      r_ger = this.r_ger;

    end % compute_r_ger()

    function r_gei = compute_r_gei(this, dNm)
    % Computes geocentric equatorial inertial position vector.
    %
    % Parameters
    %   dNm - Date number at which the position vector occurs
    %
    % Returns
    %   r_gei - Geocentric equatorial inertial position vector [er]
      if ~isequal(dNm, this.dNm)
        this.dNm = dNm;
        this.r_gei = Coordinates.ger2gei(this.r_ger, dNm);

      end % if
      r_gei = this.r_gei;

    end % compute_r_gei()

  end % methods
  
end % classdef
